spec/dockly/docker/ecr_spec.rb

```diff
@@ -0,0 +1,79 @@
+require 'spec_helper'
+
+describe Dockly::Docker::ECR do
+  subject do
+    described_class.new(name: 'ecr', server_address: registry)
+  end
+
+  let(:username) { 'AWS' }
+  let(:password) { 'password' }
+  let(:registry) { 'accountid.dkr.ecr.region.amazonaws.com' }
+  let(:endpoint) { "https://#{registry}" }
+  let(:ecr_client) { Aws::ECR::Client.new(stub_responses: true) }
+
+  after do
+    Aws.config[:ecr] = { stub_responses: {} }
+  end
+
+  describe '#authenticate!' do
+    context 'when AWS provides a username and password for ECR' do
+      before do
+        Aws.config[:ecr] = {
+          stub_responses: {
+            get_authorization_token: {
+              authorization_data: [
+                {
+                  authorization_token: Base64.encode64("#{username}:#{password}"),
+                  proxy_endpoint: endpoint
+                }
+              ]
+            }
+          }
+        }
+      end
+
+      context 'and docker auth succeeds' do
+        before do
+          allow(::Docker)
+            .to receive(:authenticate!)
+            .with({
+              'serveraddress' => endpoint,
+              'username' => username,
+              'password' => password
+            })
+        end
+
+        it 'does not error' do
+          expect { subject.authenticate! }.to_not raise_error
+        end
+      end
+
+      context 'and docker auth raises and error' do
+        before do
+          allow(::Docker)
+            .to receive(:authenticate!)
+            .and_raise(::Docker::Error::AuthenticationError)
+        end
+
+        it 'raises' do
+          expect { subject.authenticate! }.to raise_error
```

Comment:
Is it possible to check the class of Error? Would be weird if something else was raised here and the test still passed

Reply:
To follow the existing pattern in `Dockly::Docker::Registry`, I actually was just `raise`ing here, which produces a `StandardError`, so I think any error would end up as a subclass of that anyways

Reply:
Ah, yep. Well, almost any. Tom and I recently discovered that RSpec errors themselves don't subclass StandardError :D

Reply:
Spooky! Sounds like a rough day 😄 If that sort of thing is liable to get caught up in `raise_error`, certainly happy to improve the pattern here

Reply:
If it's just an error with text, I'm pretty sure you can check via: `raise_error(/text of the message/)`